tests/phpunit/CRM/Core/BAO/MessageTemplateTest.php

```diff
@@ -485,6 +485,52 @@ public function testContactTokens(): void {
     $this->assertEquals($expected_parts[2], $returned_parts[2]);
   }
 
+  /**
+   * Assuming that `testContactTokens()` has asserted tokens work en masse, we have another
+   * question -- do the tokens work the same when evaluated en-masse and individually?
+   */
+  public function testTokensIndividually() {
+    // Freeze the time at the start of the test, so checksums don't suffer from second rollovers.
+    $restoreTime = $this->useFrozenTime();
+
+    $this->hookClass->setHook('civicrm_tokenValues', [$this, 'hookTokenValues']);
+    $this->hookClass->setHook('civicrm_tokens', [$this, 'hookTokens']);
+
+    $this->createCustomGroupWithFieldsOfAllTypes([]);
+    $tokenData = $this->getOldContactTokens();
+    $this->setupContactFromTokeData($tokenData);
+
+    $ctx = ['contactId' => $tokenData['contact_id']];
+    $render = function (string $templateText) use ($ctx, $tokenData) {
+      try {
+        return CRM_Core_TokenSmarty::render(['text' => $templateText], $ctx)['text'];
+      }
+      catch (\Throwable $t) {
+        return "EXCEPTION:" . $t->getMessage();
+      }
+    };
+
+    // Build $tokenLines, a list of expressions like 'contact.display_name:{contact.display_name}'
+    $tokenLines = [];
+    $tokenNames = array_keys($this->getAdvertisedTokens());
+    foreach ($tokenNames as $tokenName) {
+      $tokenLines[] = trim($tokenName, '{}') . ':' . $tokenName;
+    }
+    foreach (array_keys($tokenData) as $key) {
+      $tokenLines[] .= "contact.$key:{contact.$key}";
+    }
+    $tokenLines = array_unique($tokenLines);
```

Comment:
In reading `testContactTokens()`, I got a bit confused because there were a few distinct authorities providing token-lists, and they don't all agree. This takes all of them (i.e the union of "advertised tokens" and "old tokens").

Reply:
Yeah advertised tokens is a subset of 'preferred tokens' - which basically = 'all the fields apiv4 can handle'. Old tokens are 'stuff we also support' - so for example `{contact.gender_id:name}` is  a 'preferred token' - but we don't present it to end users are they likely have no use for it (a better example is `{contribution.contribution_status_id:name}` which is very useful in message templates

Reply:
we test `advertisedTokens` so if something causes them to change we decide whether to let the change bubble up or not (e.g someone adds a new field & the token list changes/ test fails - then we decide if the new field should be advertised)